ugs-platform/ugs-platform-ugscore/src/main/java/com/willwinder/ugs/nbp/core/toolbars/Bundle.properties

```diff
@@ -0,0 +1,4 @@
+mainWindow.swing.firmware.toolbarTitle = Firmware selection toolbar
```

Comment:
I agree that the NetBeans documentation isn't great once you try to extend the default behavior.  There is a lot of great example-driven documentation though, which is nice (when there is an example close to what you need). The developers are pretty active and I've gotten decent support on the mailing list. [Here is a good example :)](http://wiki.netbeans.org/DevFaqActionsAddAtRuntime) 

I finally see what you were originally saying about the need for `displayName` to be a constant. I thought that the TopComponents were using derived strings, but actually, they call a `setName` method to do the same thing as the `putValue(NAME, ...)` method you found.

There are two main reasons I wanted the `ActionRegistrationService`:
1. Macros. This is the original reason. These are completely dynamic, so there is no way to register them with individual classes.
2. Repetitive actions. The `JogActionService` has almost 20 actions that sometimes differ by just a few characters. Should these be separate classes? I already had the service available for macros, so I extended it to add the default keyboard shortcuts.

I think that the stale instance objects are the most compelling reason for your first option, that alone is enough of a reason that I really like the approach you have here. The extra bundle files it seems is a non-issue (they aren't even really needed for the code in this PR). Maybe having 20 small classes with descriptive annotations is better than 20 calls to the `ActionRegistrationService` with a bunch of cryptic arguments.

It still needs to be a hybrid approach though. I don't see an alternative for Macros, and `ActionRegistrationService.createAndLocalizeFullMenu` needs to be called somewhere to localize the menu hierarchy.